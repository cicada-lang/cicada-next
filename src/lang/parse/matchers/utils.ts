import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
