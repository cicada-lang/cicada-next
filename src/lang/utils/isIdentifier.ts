const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
