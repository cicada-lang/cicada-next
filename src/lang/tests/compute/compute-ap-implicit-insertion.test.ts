import { expect, test } from "vitest"
import { runCode } from "../utils"

test("compute ApImplicit -- insertion -- id", async () => {
  const output = await runCode(`

function id(implicit T: Type, x: T): T {
  return x
}

compute id(sole)
compute id("abc")

`)

  expect(output).toMatchInlineSnapshot(`
    "sole: Trivial
    \\"abc\\": String"
  `)
})

test("compute ApImplicit -- insertion -- infer", async () => {
  const output = await runCode(`

function infer(implicit T: Type, x: T): Type {
  return T
}

compute infer(sole)
compute infer("abc")

`)

  expect(output).toMatchInlineSnapshot(`
    "Trivial: Type
    String: Type"
  `)
})

test("compute ApImplicit -- insertion -- infer2", async () => {
  const output = await runCode(`

function infer2(implicit A: Type, implicit B: Type, x: A, y: B): Type {
  return Pair(A, B)
}

compute infer2(sole, "abc")
compute infer2(sole, sole)
compute infer2("abc")(sole)
compute infer2("abc")("abc")

`)

  expect(output).toMatchInlineSnapshot(`
    "exists (Trivial) String: Type
    exists (Trivial) Trivial: Type
    exists (String) Trivial: Type
    exists (String) String: Type"
  `)
})

test("compute ApImplicit -- insertion -- infer2 -- over", async () => {
  const output = await runCode(`

function infer2(implicit A: Type, x: A, implicit B: Type, y: B): Type {
  return Pair(A, B)
}

compute infer2(sole, "abc")
compute infer2(sole, sole)
compute infer2("abc")(sole)
compute infer2("abc")("abc")

`)

  expect(output).toMatchInlineSnapshot(`
    "exists (Trivial) String: Type
    exists (Trivial) Trivial: Type
    exists (String) Trivial: Type
    exists (String) String: Type"
  `)
})

test("compute ApImplicit -- insertion -- inferReturnType", async () => {
  const output = await runCode(`

function inferReturnType(implicit A: Type, implicit B: Type, x: (A) -> B): Type {
  return B
}

function stringToTrivial(_: String): Trivial {
  return sole
}

compute inferReturnType(stringToTrivial)

function id(implicit T: Type, x: T): T {
  return x
}

compute inferReturnType(id(implicit String))

`)

  expect(output).toMatchInlineSnapshot(`
    "Trivial: Type
    String: Type"
  `)
})

test("compute ApImplicit -- insertion -- deepWalk required", async () => {
  const output = await runCode(`

function Box(T: Type): Type {
  return Pair(T, Trivial)
}

function box(implicit T: Type, x: T): Box(T) {
  return cons(x, sole)
}

compute box("abc")
compute cons("abc", sole)

`)

  expect(output).toMatchInlineSnapshot(
    `
    "cons(\\"abc\\", sole): exists (String) Trivial
    cons(\\"abc\\", sole): exists (String) Trivial"
  `,
  )
})

test("compute ApImplicit -- insertion -- nested", async () => {
  const output = await runCode(`

function Box(T: Type): Type {
  return Pair(T, Trivial)
}

function box(implicit T: Type, x: T): Box(T) {
  return cons(x, sole)
}

compute box(box("abc"))
compute box(box(box("abc")))

`)

  expect(output).toMatchInlineSnapshot(`
    "cons(cons(\\"abc\\", sole), sole): exists (exists (String) Trivial) Trivial
    cons(cons(cons(\\"abc\\", sole), sole), sole): exists (exists (exists (String) Trivial) Trivial) Trivial"
  `)
})

test("compute ApImplicit -- insertion -- during check", async () => {
  const output = await runCode(`

function idWithTrivial(implicit T: Type, solo: Trivial, x: T): T {
  return x
}


check idWithTrivial(sole): (String) -> String

let idString: (String) -> String = idWithTrivial(sole)
compute idString

`)

  expect(output).toMatchInlineSnapshot('"(_) => _: (_: String) -> String"')
})
