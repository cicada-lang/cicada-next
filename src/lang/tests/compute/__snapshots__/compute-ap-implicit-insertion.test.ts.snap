// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`compute ApImplicit -- insertion -- deepWalk required 1`] = `
"cons("abc", sole): Pair(String, Trivial)
cons("abc", sole): Pair(String, Trivial)"
`;

exports[`compute ApImplicit -- insertion -- during check 1`] = `"(_) => _: (String) -> String"`;

exports[`compute ApImplicit -- insertion -- id 1`] = `
"sole: Trivial
"abc": String"
`;

exports[`compute ApImplicit -- insertion -- infer 1`] = `
"Trivial: Type
String: Type"
`;

exports[`compute ApImplicit -- insertion -- infer2 -- over 1`] = `
"Pair(Trivial, String): Type
Pair(Trivial, Trivial): Type
Pair(String, Trivial): Type
Pair(String, String): Type"
`;

exports[`compute ApImplicit -- insertion -- infer2 1`] = `
"Pair(Trivial, String): Type
Pair(Trivial, Trivial): Type
Pair(String, Trivial): Type
Pair(String, String): Type"
`;

exports[`compute ApImplicit -- insertion -- inferReturnType 1`] = `
"Trivial: Type
String: Type"
`;

exports[`compute ApImplicit -- insertion -- nested 1`] = `
"cons(cons("abc", sole), sole): Pair(Pair(String, Trivial), Trivial)
cons(cons(cons("abc", sole), sole), sole): Pair(Pair(Pair(String, Trivial), Trivial), Trivial)"
`;
